### ISSUE TITLE
No more per job name workers?

### ISSUE THREAD

**Issue by KidkArolis** [taskforcesh/bullmq#69]
Posted this on Twitter, but moving here for more visibility.

It seems like in BullMQ you can no longer setup workers per `jobName`? That seems like a pretty large departure (and makes it tricky to migrate from `bull`). Does this mean I should be instantiating a queue per job name from now on if I want to control concurrency of each worker? Additionally, queue.add still takes jobName as first arg, but at this point it's kind of pointless (might as well put it into payload if you need to). Unless you're planning to add the per job workers back in?

This feature was really nice for grouping various job pipelines.
E.g. you might have queue X with job types kick (concurrency 1), calculate (concurrency 5), send (concurrency 20).

In other words, I would suggest to either:

* re-addi the per job name workers, since it's a nice organisational feature
* or if you think this is an unecessary complexity and should be done in the user code, I would remove the jobName from queue.add() since it's always 1-1 queue -> worker mapping. Otherwise I can already see myself calling all of the jobs "job" or "execute" or smth generic..

This put me in a bit of a pickle when I tried migrating, and I'm now migrating back to `bull`.

**Comment by KidkArolis** [comment]
Something like this would fit right in, wouldn't it?

```js
const worker1 = new Worker('Paint', fn1, { jobName: 'cars' });
const worker2 = new Worker('Paint', fn2, { jobName: 'bikes' });
```

**Comment by KidkArolis** [comment]
Or, by default:

```js
const worker = new Worker('Paint', fn)
queue.add(data)
```

and optionally

```js
const worker = new Worker('Paint', fn, { name: 'cars' })
queue.add(data, { name: 'cars' })
```

Had a look at the internals. Not sure how I'd approach implementing this, in that while it'd be easy enough to namespace all the `type` keys by jobName, it's not so easy for the queue to know what job names are available in that case. Maybe only the `wait` key should be namespaced and a new key could be introduced to track all the types that are currently waiting.

Should I just go with many queues approach :)? A queue per processing function.

**Comment by stansv** [comment]
Hello,

In previous version of Bull named processors do not work exactly as you described — you cannot create multiple workers for the same queue so each of them only process jobs he can consume. Named processors was a more convenient alternative to `switch` block inside processing callback where you select what to do depending on job type.  The only way to achieve what you want is to create multiple queues.

**Comment by manast** [comment]
@KidkArolis the name is mandatory now, mostly so that it can be used by UI:s.

**Comment by KidkArolis** [comment]
Thanks for the responses.
Naming jobs for UI makes sense - it's like a display label, not impacting functionality of the queue.

Just to get this right, @stansv, either I've been using bull 3 wrong for many years (very possible) or we're not on the same page.

I would commonly do this:

```js
const queue = new Queue('email-summary')

queue.process('kick', 1, kick)
queue.process('calculate', 2, calculate)
queue.process(send', 8, send)
queue.process('clean', 1, clean)

await queue.add('kick', {}, { repeat: { cron: '0 * * * *' } })
```

And my understanding has always been that each type of job will be processed with specified concurrency. That is - `calculate` will only be done with concurrency 2 (per server instance..) and `send` will be only done with concurrency 8 (per server instance) and so on. 🤔 

The switch statement is not equivalent, because.. I couldn't control concurrency of the individual job types.

**Comment by KidkArolis** [comment]
Reading the old docs now 😮 

![image](https://user-images.githubusercontent.com/324440/69173885-9c612200-0af8-11ea-8167-ec22964d21c0.png)

**Comment by KidkArolis** [comment]
Ok, so if you're recommendation is in this situation to create 4 queues and 4 workers (e.g. 'email-summary:kick', 'email-summary:calculate', 'email-sumary:send' and 'email-summary:clean').. then I guess I'll just switch to that and we can close this issue. Didn't realise I was using it wrong all along :(

**Comment by all43** [comment]
@stansv 
> In previous version of Bull named processors do not work exactly as you described — you cannot create multiple workers for the same queue so each of them only process jobs he can consume.

This is not true.
We do have multiple sandboxed workers for the same Q using old bull and it works exactly as described.
And it was super convenient in many cases despite leaving room for some misuse.
If there are no handler for specific name defined - job is errored, which is ok for our case.

And now we have an issue with migration to BullMQ.

Our use case:
We have many consumers (1 instance per client) and a single producer, which adds jobs to the same Q named by client. Using different queues would be difficult because:
 - Client name is dynamic, we only know client name in the moment of job creation, we will have to create new Q and destroy it every time we process new request
 - There are dozens of clients which will lead to dozens of Qs therefore making monitoring UI unusable in current state
 - In new version It is not possible to use sandboxed processors (workers) with switch statement. We could resort to using non-sandboxed processors, but it won't solve the next problem
 - Having switch statement will still pickup job not intended for specific client, so it is not a replacement for old bull named jobs.

@manast maybe we could re-open this one? Or some workaround to create dynamic Qs and monitor dozens Qs as one exist?

**Comment by manast** [comment]
The thing is that in old bull it worked exactly as a switch statement, it is just that the switch case was implicit in the queue instead of leaving the switch to you. So in order to avoid confusion it was removed from BullMQ, it is still the same functionality as before, you just need to add the switch case yourself. You can still use different names for the jobs and use the name field in your switch statement.

On the Pro version we have the concept of "groups", which basically works as "ephemeral"* micro queues that are consumed in a round robin fashion, so in some sense consuming jobs in a more fair way than when just using a standard queue. You can read more about it here: https://docs.bullmq.io/bullmq-pro/groups

* they do not exist if there are no jobs in a given group

**Comment by all43** [comment]
Thank you for your reply!
I still can't figure out how to use sandboxed processors in combination with switch statement - looking into the code it feels like it is not possible. Maybe I have to give up sandboxed processors at all - besides this issue, if the typings are correct, I can't update job data using sandboxed processor anymore.
Thank you for Pro groups suggestion - it is a bit different from our current case, but might be suitable for the task. I'll consider it.

**Comment by manast** [comment]
@all43 can you give me more details on why it is not possible using sandboxed processors?

**Comment by all43** [comment]
@manast I mean using different sandboxed processor per job name, which was possible with the old bull.
To use built-in sandboxing we need to pass to a worker processor filename as a string, which limits as to a single sandboxed processor for the queue.
If we pass function with switch statement I don't see how to use sandboxed processors without re-implementing this functionality within function.
Please correct me if I'm missing something here.

Regarding update functionality, interface `SandboxedJob` has no `update` property
https://github.com/taskforcesh/bullmq/blob/master/src/interfaces/sandboxed-job.ts
So if the typings are correct there are no easy way to update the job data within sandboxed processor.

**Comment by manast** [comment]
@all43 I do not understand why you cannot have the switch in your processor file. You can implement different cases of the switch in different modules or whatever.

You could fill a new issue about missing "update" support in the sandboxed jobs and we can fix it quickly.

**Comment by all43** [comment]
> @all43 I do not understand why you cannot have the switch in your processor file. You can implement different cases of the switch in different modules or whatever.

It is not really the same, as all the processors will run under the same pid if we simply import them inside one script without re-implementing fork/spawnChild logic inside the script. Would be nice to have possibility to reuse built-in functionality in such cases, maybe as some kind of helper class or taking object when instantiating worker.

We could have something like:
```
// for job name 'attachment-zip' we would like to use './workers/zip.ts' script
const jobNamePrefix = 'attachment:';
const proccessorMapping = {
  zip: './workers/zip.ts',
  pdf: './workers/pdf.ts',
};
const multiWorker = new Worker('jobNamePrefix', proccessorMapping);
```

> You could fill a new issue about missing "update" support in the sandboxed jobs and we can fix it quickly.
Thank you! I've created #1415

**Comment by bvkimball** [comment]
> We could have something like:
> 
> ```
> // for job named 'zip' in the attachments Q we would like to use './workers/zip.ts' script
> const queueName = 'attachments';
> const proccessorMapping = {
>   zip: './workers/zip.ts',
>   pdf: './workers/pdf.ts',
> };
> const multiWorker = new Worker('queueName', proccessorMapping);
> ```

@all43 Does this Work, or is this just request?

**Comment by all43** [comment]
> > We could have something like:
> > ```
> > // for job named 'zip' in the attachments Q we would like to use './workers/zip.ts' script
> > const queueName = 'attachments';
> > const proccessorMapping = {
> >   zip: './workers/zip.ts',
> >   pdf: './workers/pdf.ts',
> > };
> > const multiWorker = new Worker('queueName', proccessorMapping);
> > ```
> 
> @all43 Does this Work, or is this just request?

It was just a request. I ended up setting up different quests for different subtypes of task.
As I understand previous bull implementation used a condition to call different scripts for processing jobs by name after receiving jobs body, so named jobs weren't actually isolated.
After understanding how it works internally I'm no longer asking for such a feature, as you could implement similar condition on your own.
